src/lem/constrainer.rs

```diff
@@ -81,103 +150,565 @@ impl LEM {
+                };
+            }
+            AllocHashPreimage::A3(i0, i1, i2) => {
+                if let PathKind::MaybeConcrete(concrete_path) = path_kind {
+                    // if concrete_path is true, push to slots
+                    if is_concrete_path {
+                        hash_slots.hash3_alloc.push((slots.hash3_idx, i0, i1, i2));
+                        // only once per path
+                    }
+                    // concrete path implies alloc_tgt has the same value as in the current slot
+                    hash_slots.hash3_stacks.implies_stack.push((
+                        concrete_path.clone(),
+                        slots.hash3_idx,
+                        hash,
+                        tag,
+                    ));
+                // many
+                } else {
+                    hash_slots
+                        .hash3_stacks
+                        .enforce_stack
+                        .push((slots.hash3_idx, hash, tag));
+                    hash_slots.hash3_alloc.push((slots.hash3_idx, i0, i1, i2))
+                };
+            }
+            AllocHashPreimage::A4(i0, i1, i2, i3) => {
+                if let PathKind::MaybeConcrete(concrete_path) = path_kind {
+                    // if concrete_path is true, push to slots
+                    if is_concrete_path {
+                        hash_slots
+                            .hash4_alloc
+                            .push((slots.hash4_idx, i0, i1, i2, i3));
+                        // only once per path
+                    }
+                    // concrete path implies alloc_tgt has the same value as in the current slot
+                    hash_slots.hash4_stacks.implies_stack.push((
+                        concrete_path.clone(),
+                        slots.hash4_idx,
+                        hash,
+                        tag,
+                    ));
+                // many
+                } else {
+                    hash_slots
+                        .hash4_stacks
+                        .enforce_stack
+                        .push((slots.hash4_idx, hash, tag));
+                    hash_slots
+                        .hash4_alloc
+                        .push((slots.hash4_idx, i0, i1, i2, i3))
+                };
+            }
+        }
+        Ok(())
+    }
+
+    /// Here we use the implies/enforce logic to contrain tag and hash values.
+    /// We pop stacked elements in order to get virtual path information. When
+    /// many branches are possible we will use the `implies_equal` gadget to
+    /// ensure only concrete paths are implied. Otherwise, only one path exists,
+    /// therefore it must be enforced.
+    /// Hence we use two distinct stacks, one for implications and another one
+    /// for enforcements.
+    /// Beyond that, we need to deal with the possibility of receiving no tag,
+    /// since Unhash operations don't require constraining tags.
```

Comment:
I don't think I understand the Hash/Unhash asymmetry if one requires constraining tags and the other doesn't. Don't Hash/Unhash work at a lower level than tags? Shouldn't the higher-level context either handle or delegate any tag checking in the same way in both cases?

Reply:
`HashN` consumes a tag, which is the tag of the resulting pointer. That's why we constrain that the generated pointer must have that tag.

For `UnhashN`, we just need to make sure that the hash in the deconstructed pointer equals the poseidon hash of the preimage

Reply:
But why? I guess as long as you constrain the tags somewhere in all cases, it's fine. It's not obvious to me what benefit comes from handling these differently.

Reply:
This is no longer the current behavior. We're constraining the tag now